internal/v2/s2av2_e2e_test.go

```diff
@@ -335,6 +348,63 @@ func TestGRPCFallbackEndToEndUsingFakeS2AOverTCP(t *testing.T) {
 	if !fallbackCalled {
 		t.Errorf("fallbackHandshake is not called")
 	}
+	if got, want := testRetryer.Attempts(), 5; got != want {
+		t.Errorf("expecting retryer attempts count:[%v], got [%v]", want, got)
+	}
+}
+
+func TestGRPCRetryAndFallbackEndToEndUsingFakeS2AOverTCP(t *testing.T) {
+	// Set for testing only.
+	fallback.FallbackTLSConfigGRPC.InsecureSkipVerify = true
+	// Set an invalid token to trigger failures and retries when talking to S2A.
+	os.Setenv(accessTokenEnvVariable, "invalid_token")
+	testRetryer := retry.NewRetryer()
+	NewRetryer = func() *retry.S2ARetryer {
+		return testRetryer
+	}
+
+	clientS2AAddr := startFakeS2A(t, "TestE2ETCP_token")
+	grpclog.Infof("Fake handshaker for client running at address: %v", clientS2AAddr)
+	serverS2AAddr := startFakeS2A(t, "TestE2ETCP_token")
+	grpclog.Infof("Fake handshaker for server running at address: %v", serverS2AAddr)
+
+	// Start the server.
+	localIdentities := []*commonpbv1.Identity{
+		{
+			IdentityOneof: &commonpbv1.Identity_Hostname{
+				Hostname: "test_rsa_server_identity",
+			},
+		},
+	}
+	serverAddr := startServer(t, serverS2AAddr, localIdentities)
+	fallbackServerAddr := startFallbackServer(t)
+	t.Logf("server running at address: %v", serverAddr)
+	t.Logf("fallback server running at address: %v", fallbackServerAddr)
+
+	// Finally, start up the client.
```

Comment:
To make sure I understand: in this test, we retry 5 times and then fall back. Is that correct?

Is it also possible to add an e2e test where we e.g. retry once and then succeed?

Similarly for the other e2e test.

Reply:
that's right.

it's a little difficult to test the scenario of retry once and then succeed, with our fake s2a server.
Instead I was able to add a test for this logic in retry_test.go, let me know if that's okay:)